### PR TITLE
Implement Opportunistic Charging

### PR DESCRIPTION
**Pull request by r-barnes**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -113,20 +113,20 @@ follows subtly different paths which are still labeled as the same route. We
 parse the GTFS data into the following format:
 
  * `trip_id`:              A unique identifier for the trip. Taken from GTFS
- * `trip_headsign`:        The bus's public LED readout for the trip. Taken from 
+ * `trip_headsign`:        The bus's public LED readout for the trip. Taken from
```
Comment (nickolasclarke) [inline README.md:116]: this is often not present. We'll want to make this an optional field if it is required elsewhere in code.

File: src/dispatch_package/dispatch.cpp [hunk 2]
```diff
@@ -47,9 +47,9 @@ std::unordered_map<depot_id_t, int> count_buses(const trips_t &trips){
   //Load the priority queue
```
Comment (nickolasclarke) [inline src/dispatch_package/dispatch.cpp:47]: this is just tracking the buses coming and entering the depots yes?

File: src/dispatch_package/dispatch.cpp [hunk 3]
```diff
@@ -109,13 +110,14 @@ void run_block(
   };
 
   const auto end_trip = [&](trips_t::iterator trip, kilowatt_hours energy_left) -> void {
+    const seconds charge_time = std::min(params.battery_cap_kwh, params.battery_cap_kwh - energy_left)/params.depot_charger_rate;
```
Comment (nickolasclarke) [inline src/dispatch_package/dispatch.cpp:113]: this implies fully charging at a depot, as we did before, yes?

**Review by nickolasclarke — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (r-barnes) [reply README.md:116]: Okay.

File: src/dispatch_package/dispatch.cpp (hunk 2 above)
Comment (r-barnes) [reply src/dispatch_package/dispatch.cpp:47]: Yes.

File: src/dispatch_package/dispatch.cpp (hunk 3 above)
Comment (r-barnes) [reply src/dispatch_package/dispatch.cpp:113]: Yes.

**Pull request merged**